xclim/indices/_agro.py

```diff
@@ -506,6 +519,197 @@ def cool_night_index(
+
+    Returns
+    -------
+    xarray.DataArray, [mm]
+        Absolute Dryness Index values.
+
+    Notes
+    -----
+    Given that this indice only examines monthly total accumulations for six-month periods depending on the hemisphere,
+    it is possible to send in DataArrays containing only these timesteps. Users should be aware that due to the missing
+    values checks in wrapped Indicators, datasets that are missing several months will be flagged as invalid. This check
+    can be ignored by setting the following context:
+
+    .. code-block:: python
+
+        with xclim.set_options(check_missing="skip"):
+            di = dryness_index(pr, evspsblpot)
+
+    Let :math:`Wo` be the initial useful soil water reserve (typically "200 mm"), :math:`P` be precipitation,
+    :math:`T_{v}` be the potential transpiration in the vineyard, and :math:`E_{s}` be the direct evaporation from the
+    soil. Then the Dryness Index, or the estimate of soil water reserve at the end of a period (1 April to 30 September
+    in the Northern Hemispherere or 1 October to 31 March in the Southern Hemisphere), can be given by the following
+    formulae:
+
+    .. math::
+
+        W = \sum_{\text{April 1}}^{\text{September 30}} \left( Wo + P - T_{v} - E_{s} \right)
+
+    or (for the Southern Hemisphere):
+
+    .. math::
+
+        W = \sum_{\text{October 1}}^{\text{March 31}} \left( Wo + P - T_{v} - E_{s} \right)
+
+    Where :math:`T_{v}` and :math:`E_{s}` are given by the following formulae:
+
+    .. math::
+
+        T_{v} = ETP * k
+
+    and
+
+    .. math::
+
+        E_{s} = \frac{ETP}{N}\left( 1 - k \right) * JPm
+
+    Where :math:`ETP` is evapotranspiration, :math:`N` is the number of days in the given month. :math:`k` is the
+    coefficient for radiative absorption given by the vine plant architecture, and :math:`JPm` is the number of days of
+    effective evaporation from the soil per month, both provided by the following formulae:
+
+    .. math::
+
+        k = \begin{cases}
+            0.1, & \text{if month = April (NH) or October (SH)}  \\
+            0.3, & \text{if month = May (NH) or November (SH)}  \\
+            0.5, & \text{if month = June - September (NH) or December - March (SH)} \\
+            \end{cases}
+
+    .. math::
+
+        JPm = \max\left( P / 5, N \right)
+
+    Examples
+    --------
+    >>> from xclim.indices import dryness_index
+    >>> dryi = dryness_index(pr_dataset, evspsblpot_dataset, wo="200 mm")
+
+    References
+    ----------
+    :cite:cts:`tonietto_multicriteria_2004,riou_determinisme_1994`
+
+    """
+    if parse_offset(freq) != (1, "A", True, "JAN"):
+        raise ValueError(f"Freq not allowed: {freq}. Must be `YS` or `AS-JAN`")
+
+    # Resample all variables to monthly totals in mm units.
+    evspsblpot = (
+        convert_units_to(evspsblpot, "mm", context="hydro").resample(time="MS").sum()
+    )
+    pr = convert_units_to(pr, "mm", context="hydro").resample(time="MS").sum()
```

Comment:
This will only work if the variable has a correct `standard_name` (`water_potential_evaporation_flux`), otherwise `convert_units_to` will fail as the dimensionalities are wrong ([mass]/([time][length]^2)  vs [length]). In either case, `context='hydro'` does nothing as it doesn't implement this transformation.

As you can already assume correct inputs, but not the presence of a `standard_name` I suggest to perform the transformation explicitely:
```suggestion
    evspsblpot =  rate2amount(evspsblpot, out_units="mm").resample(time="MS").sum()
    pr = rate2amount(pr, out_units="mm").resample(time="MS").sum()
```

Reply:
Thanks!

Reply:
I'm running into a strange issue here. It appears that evspsblpot's units are not being recognized properly. From the ncdump:
```
        float evspsblpot(location, time) ;
                evspsblpot:_FillValue = NaNf ;
                evspsblpot:cell_methods = "time: mean within days" ;
                string evspsblpot:description = "Total potential evaporation thickness converted to mass flux using a water density of 1000 kg/m³." ;
                evspsblpot:long_name = "Mean daily potential evaporation flux" ;
                evspsblpot:original_long_name = "Potential evaporation" ;
                evspsblpot:original_variable = "pev" ;
                evspsblpot:standard_name = "water_potential_evaporation_flux" ;
                evspsblpot:units = "kg m-2 s-1" ;
                evspsblpot:coordinates = "lat lon"
```

Trace back:
```
../../core/indicator.py:845: in __call__
    outs = self.compute(**compute_das, **kwargs, **var_kwargs)
<boltons.funcutils.FunctionBuilder-54>:2: in dryness_index
    ???
../../core/units.py:976: in wrapper
    out = func(*args, **kwargs)
../../indices/_agro.py:632: in dryness_index
    evspsblpot = rate2amount(evspsblpot, out_units="mm").resample(time="MS").sum()
../../core/units.py:753: in rate2amount
    return _rate_and_amount_converter(
../../core/units.py:676: in _rate_and_amount_converter
    out = convert_units_to(out, out_units)
../../core/units.py:398: in convert_units_to
    out = source.copy(data=units.convert(source.data, source_unit, target_unit))
/home/tjs/mambaforge/envs/xclim/lib/python3.10/site-packages/pint/facets/plain/registry.py:911: in convert
    return self._convert(value, src, dst, inplace)
/home/tjs/mambaforge/envs/xclim/lib/python3.10/site-packages/pint/facets/context/registry.py:393: in _convert
    return super()._convert(value, src, dst, inplace)
/home/tjs/mambaforge/envs/xclim/lib/python3.10/site-packages/pint/facets/nonmultiplicative/registry.py:206: in _convert
    return super()._convert(value, src, dst, inplace)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pint.registry.UnitRegistry object at 0x7fa7aebca290>
value = array([[-0.13347669,  0.17290459,  0.16369012, ...,  0.07921789,
         0.03846444,  0.09547401],
       [ 0.3923424...789,  0.22684312],
       [ 0.02936513,  0.0195477 ,  0.01270458, ...,  0.00563113,
         0.00070897, -0.00178045]])
src = <UnitsContainer({'kilogram': 1, 'meter': -2})>
dst = <UnitsContainer({'millimeter': 1})>, inplace = False
check_dimensionality = True

~~~

>               raise DimensionalityError(src, dst, src_dim, dst_dim)
E               pint.errors.DimensionalityError: Cannot convert from 'kilogram / meter ** 2' ([mass] / [length] ** 2) to 'millimeter' ([length])
```

Why isn't `s-1` being found?

Reply:
oups, my bad here. `rate2amount` is converting the kg m-2 s-1 into kg m-2, and then trying to convert to mm. This is a transformation that needs the `hydro` context turned on, but the  `rate2amount` function doesn't pass it.

```python3
evspsblpot =  convert_units_to(rate2amount(evspsblpot), "mm", context="hydro").resample(time="MS").sum()
pr = convert_units_to(rate2amount(pr), "mm", context="hydro").resample(time="MS").sum()
```
or (but I'm less sure here):
```python3
with units.context("hydro"):
    evspsblpot =  rate2amount(evspsblpot, out_units="mm").resample(time="MS").sum()
    pr = rate2amount(pr, out_units="mm").resample(time="MS").sum()
```

Reply:
Or, maybe even better as it is completely explicit:
```python3
evspsblpot =  amount2lwethickness(rate2amount(evspsblpot), out_units="mm").resample(time="MS").sum()
pr =   amount2lwethickness(rate2amount(pr), out_units="mm").resample(time="MS").sum()
```

Reply:
I'd opt for the first option (`Explicit is better than implicit.`). Will push changes to the function itself.